crates/re_data_source/src/data_loader/mod.rs

```diff
@@ -0,0 +1,175 @@
+use std::sync::Arc;
+
+use once_cell::sync::Lazy;
+
+use re_log_types::{ArrowMsg, DataRow, LogMsg};
+
+// ---
+
+/// A [`DataLoader`] loads data from a file path and/or a file's contents.
+///
+/// Files can be loaded in 3 different ways:
+/// - via the Rerun CLI (`rerun myfile.jpeg`),
+/// - using drag-and-drop,
+/// - using the open dialog in the Rerun Viewer.
+///
+/// All these file loading methods support loading a single file, many files at once, or even
+/// folders.
+/// ⚠ Drag-and-drop of folders does not yet work on the web version of Rerun Viewer ⚠
+///
+/// Rerun comes with a few [`DataLoader`]s by default:
+/// - [`RrdLoader`] for `.rrd` files,
+/// - [`ArchetypeLoader`] for:
+///     - 3D models (`.gltf`, `.glb`, `.obj`)
+///     - Images (most mainstream image formats)
+///
+/// ## Execution
+///
+/// **All** registered [`DataLoader`]s get called when a user tries to open a file, unconditionally.
+/// This gives [`DataLoader`]s maximum flexibility to decide what files they are interested in, as
+/// opposed to e.g. only being able to look at files' extensions.
+///
+/// On native, [`DataLoader`]s are executed in parallel.
+//
+// TODO: issue -> DataLoader should support any URI, any platform, etc
+// TODO: issue -> dragndrop folder on web
+// TODO: issue -> make it available to SDKs
+pub trait DataLoader: Send + Sync {
+    /// Name of the [`DataLoader`].
+    ///
+    /// Doesn't need to be unique.
+    fn name(&self) -> String;
+
+    /// Loads data from a file on the local filesystem and sends it to `tx`.
+    ///
+    /// This is generally called when opening files with the Rerun CLI or via the open menu in the
+    /// Rerun Viewer on native platforms.
+    ///
+    /// The passed-in `store_id` is a shared recording created by the file loading machinery:
+    /// implementers can decide to use it or not (e.g. it might make sense to log all images with a
+    /// similar name in a shared recording, while an rrd file is already its own recording).
+    ///
+    /// `path` isn't necessarily a _file_ path, but can be a directory as well: implementers are
+    /// free to handle that however they decide.
+    ///
+    /// ## Error handling
+    ///
+    /// Most implementers of `load_from_file` are expected to be asynchronous in nature.
+    ///
+    /// Asynchronous implementers should make sure to fail early (and thus synchronously) when
+    /// possible (e.g. didn't even manage to open the file).
+    /// Otherwise, they should log errors that happen in an asynchronous context.
+    ///
+    /// If a [`DataLoader`] has no interest in the given file, it should successfully return
+    /// without pushing any data into `tx`.
+    #[cfg(not(target_arch = "wasm32"))]
+    fn load_from_file(
+        &self,
+        store_id: re_log_types::StoreId,
+        path: std::path::PathBuf,
+        tx: std::sync::mpsc::Sender<LoadedData>,
+    ) -> Result<(), DataLoaderError>;
```

Comment:
I don't want to go there because the former accepts folder paths while the latter doesn't, and I don't want to do some automagic filtering of folders on behalf of the user...